packages/tpetra/core/src/Tpetra_CrsSingletonFilter_LinearProblem_def.hpp

```diff
@@ -0,0 +1,919 @@
+      // Check to see if this column already eliminated by the row check above
+      if (RowMapColors_Data(i2, 0) != 1) {
+        RowHasColWithSingletonData(i2, 0)++;  // Increment col singleton counter for ith row
+        RowMapColors_Data(i2, 0) = 2;         // Use 2 for now, to distinguish between row eliminated directly or via column singletons
+        ColMapColors_Data(j, 0)  = 1;
+        localNumSingletonCols_++;
+        // If we delete a row, we need to keep track of associated column entries that were also deleted
+        // in case all entries in a column are eventually deleted, in which case the column should
+        // also be deleted.
+        GetRow(i2, NumIndices, localIndices);
+        for (size_t jj = 0; jj < NumIndices; jj++) {
+          NewColProfilesData(localIndices[jj], 0)--;
+        }
+      }
+    }
+    // Check if some other processor eliminated this column
+    else if (ColHasRowWithSingletonData(j, 0) == 1 && RowMapColors_Data(i2, 0) != 1) {
+      ColMapColors_Data(j, 0) = 1;
+    }
+  }
+
+  TEUCHOS_TEST_FOR_EXCEPTION_CLASS_FUNC(RowHasColWithSingleton.normInf() > 1, std::runtime_error,
+                                        "At least one row is associated with two singleton columns, can't handle it.");
+
+  // Generate arrays that keep track of column singleton row, col and pivot info needed for post-solve phase
+  CreatePostSolveArrays(localRowIDofSingletonCol, ColProfiles, NewColProfiles, ColHasRowWithSingleton);
+
+  for (local_ordinal_type i = 0; i < localNumRows; i++) {
+    if (RowMapColors_Data(i, 0) == 2) RowMapColors_Data(i, 0) = 1;  // Convert all eliminated rows to same color
+  }
+
+  const Teuchos::Ptr<local_ordinal_type> gRowsPtr(&globalNumSingletonRows_);
+  const Teuchos::Ptr<local_ordinal_type> gColsPtr(&globalNumSingletonCols_);
+  auto rowComm = FullMatrix()->getRowMap()->getComm();
+  Teuchos::reduceAll<local_ordinal_type, local_ordinal_type>(*rowComm, Teuchos::REDUCE_SUM, localNumSingletonRows_, gRowsPtr);
+  Teuchos::reduceAll<local_ordinal_type, local_ordinal_type>(*rowComm, Teuchos::REDUCE_SUM, localNumSingletonCols_, gColsPtr);
+
+  AnalysisDone_ = true;
+  return;
+}
+
+//==============================================================================
+
+template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+Teuchos::RCP<const Tpetra::Map<LocalOrdinal, GlobalOrdinal, Node>>
+CrsSingletonFilter_LinearProblem<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+    GenerateReducedMap(const Teuchos::RCP<const map_type>& originalMap,
+                       const Teuchos::RCP<vector_type_int>& mapColors, int color) {
+  // Vector to hold the reduced global indices
+  std::vector<GlobalOrdinal> myReducedGlobalIndices;
+
+  // Get the global IDs owned by the calling process
+  auto myGlobalIndices = originalMap->getMyGlobalIndices();
+
+  // Iterate through the global indices to find the ones that match the color
+  for (size_t i = 0; i < myGlobalIndices.size(); i++) {
```

Comment:
One `parallel_reduce` to determine number of map indices, then memory allocation and fill using `parallel_for`.